### PR TITLE
[Yaml] Remove escaping regex

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

This PR replaces regexs in `Escaper` by call to `strcspn`. It is much easier to read imo and performance aren't degraded.
I also updated the data provider of a test because it was in practice testing nothing (it is checking if values are quoted as expected but as they always began with `\t` it was always quoted anyway).

Edit: quotes are moved to indicators as they are supported inside plain scalars as said by @stefk:

> It seems those characters are perfectly valid in unquoted strings:
> 
> http://yaml.org/spec/1.2/spec.html#id2788859
> http://yaml-online-parser.appspot.com/?yaml=foo%3A+ba%22r%0Abaz%3A+q%27u%27x%27%0A&type=canonical_yaml

### CODE REVIEW

File: src/Symfony/Component/Yaml/Escaper.php [hunk 1]
```diff
@@ -48,7 +52,7 @@ class Escaper
      */
     public static function requiresDoubleQuoting($value)
     {
-        return preg_match('/'.self::REGEX_CHARACTER_TO_ESCAPE.'/u', $value);
+        return strlen($value) !== strcspn($value, implode('', self::$escapees));
```
Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Escaper.php]: This would force escaping values containing `"`, `/` and `\` that weren't escaped before. Does it matter ?

File: benchmark.php [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?php
```
Comment (nicolas-grekas) [inline benchmark.php]: not sure about this file :)

File: src/Symfony/Component/Yaml/Escaper.php [hunk 3]
```diff
@@ -48,7 +52,10 @@ class Escaper
      */
     public static function requiresDoubleQuoting($value)
     {
-        return preg_match('/'.self::REGEX_CHARACTER_TO_ESCAPE.'/u', $value);
+        // Removes quotes and slashes
+        $mask = implode('', array_slice(self::$escapees, 5));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Escaper.php]: we'd better create a new const for this, and deprecate REGEX_CHARACTER_TO_ESCAPE meanwhile

File: src/Symfony/Component/Yaml/Escaper.php [hunk 4]
```diff
@@ -78,9 +85,13 @@ public static function requiresSingleQuoting($value)
             return true;
         }
 
-        // Determines if the PHP value contains any single characters that would
-        // cause it to require single quoting in YAML.
-        return preg_match('/[ \s \' " \: \{ \} \[ \] , & \* \# \?] | \A[ \- ? | < > = ! % @ ` ]/x', $value);
+        // First character is reserved
+        if ($value && in_array($value[0], array('-', '|', '<', '>', '=', '!', '%', '@', '`'))) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Escaper.php]: can't we use another const here with strspn?

Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Escaper.php]: (maybe not for the const, strspn still?)

Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Escaper.php]: Not sure about the perfs but i'll try !

File: benchmark.php (hunk 2 above)
Comment (Ener-Getick) [inline benchmark.php]: Indeed, i should stop using `-A` ^^

File: src/Symfony/Component/Yaml/Escaper.php (hunk 3 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Escaper.php]: 👍 for the const
But i don't think we need to deprecate this constant, it is internal in 3.0 anyway so let's just remove it here ?

File: src/Symfony/Component/Yaml/Escaper.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Escaper.php]: ok if it's internal in 3.0 then fine!

Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Escaper.php]: I think `strspn` is slower but it is clearer, what do you prefer?

File: src/Symfony/Component/Yaml/Escaper.php [hunk 5]
```diff
@@ -21,23 +21,31 @@ class Escaper
...
-                                     "\x18",  "\x19",  "\x1a",  "\x1b",  "\x1c",  "\x1d",  "\x1e",  "\x1f",
-                                     "\xc2\x85", "\xc2\xa0", "\xe2\x80\xa8", "\xe2\x80\xa9");
-    private static $escaped = array('\\\\', '\\"', '\\\\', '\\"',
-                                     '\\0',   '\\x01', '\\x02', '\\x03', '\\x04', '\\x05', '\\x06', '\\a',
-                                     '\\b',   '\\t',   '\\n',   '\\v',   '\\f',   '\\r',   '\\x0e', '\\x0f',
-                                     '\\x10', '\\x11', '\\x12', '\\x13', '\\x14', '\\x15', '\\x16', '\\x17',
-                                     '\\x18', '\\x19', '\\x1a', '\\e',   '\\x1c', '\\x1d', '\\x1e', '\\x1f',
-                                     '\\N', '\\_', '\\L', '\\P');
+    private static $escapees = array(
+        '\\', '\\\\', '\\"', '"', '/',
```
Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Escaper.php]: BTW this fixes a "bug", `/` was missing.

**Comment by fabpot** [comment]
As it's not a bug fix, this should be done in master.

**Comment by Ener-Getick** [comment]
ok, retargeted and rebased. 
So no refactoring won't be merged in older branches? That's probably a good thing :)

**Comment by fabpot** [comment]
That's the main idea, refactorings do not fix anything, so they should be done in master (the only exception is when it gives a huge performance improvement and the changes are trivial).

File: src/Symfony/Component/Yaml/Escaper.php (hunk 5 above)
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Escaper.php]: Shouldn't this be done in 2.7 then?

File: src/Symfony/Component/Yaml/Inline.php [hunk 6]
```diff
@@ -684,7 +684,7 @@ public static function evaluateBinaryScalar($scalar)
 
     private static function isBinaryString($value)
     {
-        return !preg_match('//u', $value) || preg_match('/[^\x09-\x0d\x20-\xff]/', $value);
+        return !preg_match('//u', $value) || preg_match('/[^\x00-\x1f\x20-\xff]/', $value);
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: What's the reason for this change?

Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Inline.php]: Because these caracters are escapable in double quoted strings

File: src/Symfony/Component/Yaml/Escaper.php (hunk 5 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Escaper.php]: It works in 2.7, it just that it should be escaped but that's not mandatory.

File: src/Symfony/Component/Yaml/Inline.php (hunk 6 above)
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: Ah okay, maybe we should add a comment with some reference to an external source or something like that to help understanding the code.

Comment (Ener-Getick) [inline src/Symfony/Component/Yaml/Inline.php]: in fact i'm not even sure we should use `!!binary` when dumping, most (if not all?) characters are supported in double-quoted anyway.
I'll try to find what the spec says about that.

**Comment by nicolas-grekas** [comment]
Well, since you changed some char ragnges, this looks like a bug fix to me. About the const, removing it is a BC break.
Thus, I think I'd prefer keeping the regexp and focus on the other changes for 2.7.

**Comment by stof** [comment]
> About the const, removing it is a BC break.

@nicolas-grekas the whole class is marked as internal, so not covered by the BC promise

File: src/Symfony/Component/Yaml/Escaper.php [hunk 7]
```diff
@@ -80,9 +88,14 @@ public static function requiresSingleQuoting($value)
             return true;
         }
 
-        // Determines if the PHP value contains any single characters that would
-        // cause it to require single quoting in YAML.
-        return preg_match('/[ \s \' " \: \{ \} \[ \] , & \* \# \?] | \A[ \- ? | < > = ! % @ ` ]/x', $value);
+        // First character is an indicator
+        // @see http://yaml.org/spec/1.2/spec.html#c-indicator
+        if ($value && 1 === strspn($value[0], '-?&*!|<>\'"=%@`')) {
```
Comment (stof) [inline src/Symfony/Component/Yaml/Escaper.php]: this is broken for the empty string (btw, the simpler fix for that is probably to add the empty string in the array of values above)

File: src/Symfony/Component/Yaml/Tests/DumperTest.php [hunk 8]
```diff
@@ -257,23 +257,24 @@ public function testEscapedEscapeSequencesInQuotedScalar($input, $expected)
...
+            'vertical-tab' => array("\v", '"\\v"'),
+            'form-feed' => array("\xC", '"\\f"'),
+            'carriage-return' => array("\r", '"\\r"'),
+            'escape' => array("\x1B", '"\\e"'),
+            'space' => array(' ', "' '"),
+            'double-quote' => array('"', "'\"'"),
+            'slash' => array('/', '/'),
+            'backslash' => array('\\', '\\'),
+            'next-line' => array("\xC2\x85", '"\\N"'),
+            'non-breaking-space' => array('�', '�'),
```
Comment (stof) [inline src/Symfony/Component/Yaml/Tests/DumperTest.php]: shouldn't this be ``"\_"`` in YAML ?

Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Tests/DumperTest.php]: Sure, the test is wrong.

File: src/Symfony/Component/Yaml/Escaper.php (hunk 7 above)
Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Escaper.php]: Indeed, good catch!

File: src/Symfony/Component/Yaml/Escaper.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Yaml/Escaper.php]: so why doing it then?

Comment (GuilhemN) [inline src/Symfony/Component/Yaml/Escaper.php]: Actually that's not clear what we should do here. [Non-printable characters must be escaped](http://yaml.org/spec/1.2/spec.html#id2776092) but `/` is a [printable character](http://yaml.org/spec/1.2/spec.html#nb-json) ([the spec says `\/` is available for json compatibily](http://yaml.org/spec/1.2/spec.html#id2776092)), so I guess it's up to us to decide what to do.

For readability `/` should indeed be better while `\/` is more compliant with json. If you prefer `/` I'll revert this change.

**Comment by nicolas-grekas** [comment]
So, to help moving this one forward, I'm personnaly :-1: on this:

> This PR replaces regexs in Escaper by call to strcspn. It is much easier to read imo and performance aren't degraded.

This is subjective, with no clear technical benefit, and will create merge conflicts that are thus not worth it.

**Comment by GuilhemN** [comment]
@nicolas-grekas this PR mostly intents to improve readability and maintainability (do you think it's obvious what `preg_match('/[ \s \' " \: \{ \} \[ \] , & \* \# \?] | \A[ \- ? | < > = ! % @ ` ]/x', $value);` does ?).
About merge conflicts I understand your concern but this file is [not modified often](https://github.com/symfony/symfony/blame/master/src/Symfony/Component/Yaml/Escaper.php) (last modification of a line modified here was in january 2015) so I don't think it will be a problem.

It also improves a bit scalars dumping when quotes are in the middle of the scalar (e.g. `foo'bar` was dumped as `'foo''bar'` now it is dumped as `foo'bar` as it is a valid plain scalar).

Anyway, we should at least update [the tests](https://github.com/symfony/symfony/pull/19782/files#diff-b08eb9d62892e494cadd13feea3c48d4R260) which are currently useless.

**Comment by fabpot** [comment]
Closing as I agree with @nicolas-grekas here. @GuilhemN Can you submit a PR for the tests though?
